File: microservices/simple/emailNode/src/amqp_setup.js
const amqp = require('amqplib');

const hostname = process.env.RABBITMQ_HOST || 'rabbitmq';
const port = process.env.RABBITMQ_PORT || 5672;
const exchangeName = 'email_exchange';
const exchangeType = 'topic';
const queueName = 'email_queue';

let connection = null;
let channel = null;
let setupPromise = null;

async function connect() {
  const nextConnection = await amqp.connect(`amqp://${hostname}:${port}`, { heartbeat: 60 });
  nextConnection.on('close', () => {
    connection = null;
    channel = null;
    setupPromise = null;
  });
  nextConnection.on('error', (error) => {
    console.error('RabbitMQ connection error:', error.message);
  });

  const nextChannel = await nextConnection.createChannel();
  await nextChannel.assertExchange(exchangeName, exchangeType, { durable: true });
  await nextChannel.assertQueue(queueName, { durable: true });
  await nextChannel.bindQueue(queueName, exchangeName, '*.email');

  connection = nextConnection;
  channel = nextChannel;
  return channel;
}

async function checkSetup() {
  if (connection && channel) {
    return channel;
  }

  if (!setupPromise) {
    setupPromise = connect().catch((error) => {
      setupPromise = null;
      throw error;
    });
  }

  return setupPromise;
}

module.exports = {
  checkSetup,
  get channel() {
    return channel;
  },
};

File: microservices/complex/process_order/src/amqp_setup.py
from os import environ

import pika


hostname = environ.get("rabbit_host") or environ.get("rabbit-host") or "rabbitmq"
port = int(environ.get("rabbit_port", 5672))
exchangename = "email_exchange"
exchangetype = "topic"
queue_name = "email_queue"
connection = None
channel = None


def check_setup():
    global connection, channel

    if connection is None or connection.is_closed:
        connection = pika.BlockingConnection(
            pika.ConnectionParameters(
                host=hostname,
                port=port,
                heartbeat=3600,
                blocked_connection_timeout=3600,
            )
        )

    if channel is None or channel.is_closed:
        channel = connection.channel()
        channel.exchange_declare(
            exchange=exchangename,
            exchange_type=exchangetype,
            durable=True,
        )
        channel.queue_declare(queue=queue_name, durable=True)
        channel.queue_bind(
            exchange=exchangename,
            queue=queue_name,
            routing_key="*.email",
        )
